### PR TITLE
Feature lpdev 152 create transfer request controller

### PR DESCRIPTION
**Pull request by AntonSivov**

### CODE REVIEW

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 1]
```diff
@@ -19,18 +21,21 @@ public class BtcTransferRequestService : ITransferRequestService
...
-            _bitcointApiClient = bitcointApiClient;
-            _walletClient = walletClient;
-            _log = log;
+            _transferRepository = transferRepository ?? throw new ArgumentNullException(nameof(transferRepository));
+            _walletRepository = walletRepository ?? throw new ArgumentNullException(nameof(walletRepository));
+            _bitcointApiClient = bitcointApiClient ?? throw new ArgumentNullException(nameof(bitcointApiClient));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
         }
         #endregion
         #region public
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: Do we really need initialize SourceAmount array here if you already passing wallets inside CalculateSources?

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 2]
```diff
@@ -61,7 +66,7 @@ public BtcTransferRequestService(ITransferRepository transferRepository, IWallet
                 transferRequest.DestinationAddress
             };
 
-            var isAddressesValid = await CheckAddressesValid(transferWallets);
+            var isAddressesValid = CheckAddressesValid(transferWallets);
             if (!isAddressesValid)
             {
                 return new TransferDto
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: I'd make TransferDto factory to create this kind of objects and not to pollute the BL code.

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 3]
```diff
@@ -186,6 +191,8 @@ public BtcTransferRequestService(ITransferRepository transferRepository, IWallet
 
         private async Task<ITransferInfo> CreateBtcTransfer(List<ISourceAmount> sources, string destination)
         {
+            sources = sources.Where(s=>s.Amount != 0).ToList();
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: Wouldn't it be better to use s.Amount > 0?

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 4]
```diff
@@ -195,21 +202,28 @@ public BtcTransferRequestService(ITransferRepository transferRepository, IWallet
...
                 Destination = destination,
                 Sources = (from s in sources
-                           select new ToOneAddress(s.SourceAddress, s.Amount)).ToList()
+                    select new ToOneAddress(s.SourceAddress, s.Amount)).ToList()
             };
-            var r = await _bitcointApiClient.ApiTransactionMultipletransferPostWithHttpMessagesAsync();
-            var resData = r?.Body as TransactionIdAndHashResponse;
-            if (resData?.Hash == null)
+
+            if (sources.Count != 0)
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: sources.Any()

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 5]
```diff
@@ -195,21 +202,28 @@ public BtcTransferRequestService(ITransferRepository transferRepository, IWallet
...
+            }
+           
+          
+            if (request == null || request.HasError)
             {
-                var errorCode = (r?.Body as ApiException)?.Error.Code;
-                var errorMessage = (r?.Body as ApiException)?.Error.Message;
-                if (resData == null && errorCode == "3")
+               
+                if (request != null && request.Error.Code == "3")
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: request?.Error?.Code

Comment (tarurar) [reply src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: Why hardcoding error codes?

File: src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs [hunk 6]
```diff
@@ -298,14 +312,13 @@ private List<ISourceAmount> CalculateSources(IEnumerable<ISourceAmount> sources,
...
+        private bool CheckAddressesValid(IReadOnlyList<string> addresses)
         {
             try
             {
-                foreach (var a in addresses)
+                foreach (var address in addresses)
                 {
-                    var tAd = BitcoinAddress.Create(a); //TODO Check if we need to set Network
-                    await _walletClient.CreateAddress(tAd);
+                    BitcoinAddress.Create(address);
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs]: What's the point of catching exception and returning false?

File: src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+        /// <summary>
+        /// Update transfer status.
+        /// </summary>
+        /// <param name="model">Transfer model.</param>
+        /// <returns>The Transfer Info.</returns>
+        /// <response code="200">The Transfer Info.</response>
+        /// <response code="400">Invalid model.</response>
+        [HttpPost]
+        [Route("transfers/updateStatus")]
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs]: Why not to make "transfers" route the part of controller route?

File: src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs [hunk 8]
```diff
@@ -1,11 +1,207 @@
...
+using Lykke.Service.PayInternal.Extensions;
+using Lykke.Service.PayInternal.Models;
 using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.SwaggerGen;
 
 namespace Lykke.Service.PayInternal.Controllers
 {
-    public class TransferRequestController: Controller
+    [Route("api")]
+    public class TransferRequestController : Controller
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs]: Why do we need to separate Report and Request controllers?

File: src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs [hunk 9]
```diff
@@ -1,11 +1,207 @@
...
+
+           
+            decimal dAmount;
+            if (!decimal.TryParse(amount, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out dAmount))
+            {
+                return BadRequest(ErrorResponse.Create("Amount is not a number"));
+            }
+
+            if (string.IsNullOrEmpty(sourceAddress))
+                return BadRequest(ErrorResponse.Create("Source Address doesn't exist"));
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs]: Source address is empty?

File: src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs [hunk 10]
```diff
@@ -1,11 +1,207 @@
...
+                return BadRequest(new ErrorResponse().AddErrors(ModelState));
+
+           
+            decimal dAmount;
+            if (!decimal.TryParse(amount, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out dAmount))
+            {
+                return BadRequest(ErrorResponse.Create("Amount is not a number"));
+            }
+
+            if (sourceAddressesList == null || sourceAddressesList.Count == 0 ||
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs]: sourceAddressesList .Any()

File: src/Lykke.Service.PayInternal/Models/TransferSourcesRequestModel.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using Lykke.Service.PayInternal.Core.Domain.Transfer;
+
+namespace Lykke.Service.PayInternal.Models
+{
+    public class TransferSourcesRequestModel : TransferRequestModel, ISourcesTransferRequest
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Models/TransferSourcesRequestModel.cs]: I'd suggest name it IMultipleSourcesTransferRequest

File: src/Lykke.Service.PayInternal/Modules/ServiceModule.cs [hunk 12]
```diff
@@ -79,6 +82,10 @@ private void RegisterAzureRepositories(ContainerBuilder builder)
                 AzureTableStorage<BlockchainTransactionEntity>.Create(
                     _dbSettings.ConnectionString(x => x.MerchantConnString),
                     "MerchantWalletTransactions", _log)));
+
+            builder.RegisterInstance<ITransferRepository>(new TransferRepository(
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Modules/ServiceModule.cs]: For repository you can register type, but not instance. Better register instance of INoSQLTableStorage

File: src/Lykke.Service.PayInternal/Modules/ServiceModule.cs [hunk 13]
```diff
@@ -114,6 +124,9 @@ private void RegisterServiceClients(ContainerBuilder builder)
             builder.RegisterType<LykkeMarketProfile>()
                 .As<ILykkeMarketProfile>()
                 .WithParameter("baseUri", new Uri(_settings.CurrentValue.MarketProfileServiceClient.ServiceUrl));
+            var ninjaServiceClient = new QBitNinjaClient(_settings.CurrentValue.NinjaServiceClient.ServiceUrl);
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Modules/ServiceModule.cs]: Better not to construct object yourself. Delegate object construction to autofac and just register the type.

File: src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransactionRequest.cs [hunk 14]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+using Lykke.Service.PayInternal.Core.Domain.Transfer;
+
+namespace Lykke.Service.PayInternal.AzureRepositories.Transfer
+{
+    public class TransactionRequest : ITransactionRequest
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransactionRequest.cs]: Why not TransferRequest?

File: src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransferRepository.cs [hunk 15]
```diff
@@ -17,39 +13,73 @@ public TransferRepository(INoSQLTableStorage<TransferEntity> tableStorage)
...
             {
-                return null;
+                try
+                {
+                    await _tableStorage.InsertOrMergeAsync(transfer);
+
+                }
+                catch
+                {
+                    return null;
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransferRepository.cs]: Is it expected behaviour?

File: src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        [Route("updateStatus")]
+        [SwaggerOperation("TransferUpdateStatus")]
+        [ProducesResponseType(typeof(TransferModel), (int)HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> UpdateTransferStatusAsync([FromBody] UpdateTransferStatusModel model)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(new ErrorResponse().AddErrors(ModelState));
+
+            if (string.IsNullOrEmpty(model.TransferId))
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs]: Why not to add [Required] attribute to this property and catch this case with previous piece of code? (ModelState.IsValid)

File: src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+                return BadRequest(ErrorResponse.Create("Transfer id doesn't exist"));
+
+            if (string.IsNullOrEmpty(model.TransactionHash))
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Controllers/TransferReportController.cs]: Why not to add [Required] attribute to this property and catch this case with previous piece of code? (ModelState.IsValid)

File: src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs (hunk 9 above)
Comment (tarurar) [reply src/Lykke.Service.PayInternal/Controllers/TransferRequestController.cs]: Up

File: src/Lykke.Service.PayInternal/Modules/ServiceModule.cs [hunk 18]
```diff
@@ -96,6 +103,9 @@ private void RegisterAppServices(ContainerBuilder builder)
             builder.RegisterType<MerchantWalletsService>()
                 .As<IMerchantWalletsService>();
 
+            builder.RegisterType<BtcTransferRequestService>()
+                .As<ITransferRequestService>();
```
Comment (tarurar) [inline src/Lykke.Service.PayInternal/Modules/ServiceModule.cs]: SingleInstance?

File: src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransferRepository.cs (hunk 15 above)
Comment (AntonSivov) [reply src/Lykke.Service.PayInternal.AzureRepositories/Transfer/TransferRepository.cs]: Yep

**Pull request closed without merging**